### PR TITLE
Add cookie forwarding + ssl/tls support

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Forward cookie to server and set-cookie to client.
add flag to use tls/ssl

## Changes

- Fulfill purpose.
- Carry metadata from express server to grpc client and back. 

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by fdibat — APPROVED** [review]
LGTM, but please add also a developer reviewer

File: src/server.ts [hunk 1]
```diff
@@ -33,6 +37,14 @@ export default (
         const request = req.body;
         logger.info('Received a request', { correlationId });
 
+        const cookie = req.header('cookie');
+        const clientMetadata = new Metadata();
+        if (cookie) {
+            logger.debug('cookie', cookie, { correlationId });
```
Comment (bisgardo) [inline src/server.ts:43]: In the [doc](https://concordium.atlassian.net/wiki/spaces/DO/pages/1041203217/Wallets+backend+refactoring) it says that it should be on "trace" level. I guess this translates to `verbose` here?

File: src/server.ts [hunk 2]
```diff
@@ -52,8 +64,20 @@ export default (
             }
 
             if (response) {
+                const { result, ...rest } = response;
+                const actualResponse = { ...rest, result: result.result };
+
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                const setCookie = (result as ResultAndMetadata<any>).metadata
+                    .get('set-cookie')
+                    .map((x) => x.toString());
```
Comment (bisgardo) [inline src/server.ts:73]: Add explanatory comment(s).

File: src/server.ts [hunk 3]
```diff
@@ -52,8 +64,20 @@ export default (
...
+                if (setCookie.length) {
+                    logger.debug('set-cookie:', setCookie);
```
Comment (bisgardo) [inline src/server.ts:75]: Same as above.

File: src/logger.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import winston from 'winston';
 import minimist from 'minimist';
 
 const logger = winston.createLogger({
-    level: 'info',
+    level: minimist(process.argv.slice(2)).log || 'info',
```
[the remark this answers is not in the document]
Comment (bisgardo) [reply src/logger.ts:5]: You could also move this to `app.ts` where `argv` is already parsed. Would be fine with me.

**Review by bisgardo — COMMENTED** [review]
Could you add some explanation of this metadata object? I.e. what it is/contains in relation to express and grpc.

File: src/logger.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (shjortConcordium) [reply src/logger.ts:5]: I'll move the cli args into a `config.ts`.

File: src/client.ts [hunk 5]
```diff
@@ -41,26 +43,39 @@ export default class NodeClient {
...
 
-                return command.bind(this.client)(
+                const clientMetadata = this.metadata.clone();
+                clientMetadata.merge(metadata);
+                let serverMetadata: Metadata;
+                const call: SurfaceCall = command.bind(this.client)(
                     input,
-                    this.metadata,
+                    clientMetadata,
                     // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
[the remark this answers is not in the document]
Comment (shjortConcordium) [reply src/client.ts:65]: I generally still want the rule, I want it using `any` to be annoying.

File: src/server.ts (hunk 1 above)
Comment (shjortConcordium) [reply src/server.ts:43]: Hmm, I 'translated' trace to debug. :thinking:
(I'd think that verbose < debug < trace)

**Comment by shjortConcordium** [comment]
> Could you add some explanation of this metadata object? I.e. what it is/contains in relation to express and grpc.

I've tried to add a bit. I'm not sure though, the metadata is just the grpc metadata, and I've kept it generic everywhere except in server.ts where we put cookie in and extract set-cookie. (Which is where I've added comments)

File: src/logger.ts [hunk 6]
```diff
@@ -1,16 +1,15 @@
...
+    level: logLevel || 'info',
     format: winston.format.combine(
         winston.format.timestamp(),
         winston.format.json()
     ),
     transports: [
         new winston.transports.File({
-            filename:
-                minimist(process.argv.slice(2)).logLocation || 'combined.log',
+            filename: logLocation || 'combined.log',
```
Comment (bisgardo) [inline src/logger.ts:12]: Maybe not for this PR but I think the default should be stdout.

File: src/server.ts (hunk 1 above)
Comment (bisgardo) [reply src/server.ts:45]: Right, the order is
```
const levels = {
  error: 0,
  warn: 1,
  info: 2,
  http: 3,
  verbose: 4,
  debug: 5,
  silly: 6
};
```

Using `silly` sounds silly so let's just keep it like this ;)

**Review by bisgardo — APPROVED** [review]

File: src/logger.ts (hunk 6 above)
Comment (shjortConcordium) [reply src/logger.ts]: I changed it to be so, the combined.log default was strange.

**Pull request merged**